cupy/_core/_kernel.pyx

```diff
@@ -73,12 +74,23 @@ cdef inline int _get_kind_score(int kind):
 
 
 @cython.profile(False)
-cdef inline _check_array_device_id(ndarray arr, int device_id):
-    if arr.data.device_id != device_id:
+cdef inline _check_peer_access(ndarray arr, int device_id):
+    if arr.data.device_id == device_id:
+        return
+
+    msg = (
+        f'The device where the array resides ({arr.data.device_id}) is '
+        f'different from the current device ({device_id}).'
+    )
+
+    cdef bint peer_access = device._enable_peer_access(
+        device_id, arr.data.device_id)
+    if not peer_access:
         raise ValueError(
-            'Array device must be same as the current '
-            'device: array device = %d while current = %d'
-            % (arr.data.device_id, device_id))
+            '{msg} Peer access could not be activated automatically.')
+    warnings.warn(
+        f'{msg} Peer access has been activated automatically.',
```

Comment:
I am not sure if we should enable some kind of env var to disable this automatic behavior.

Reply:
I think visible warnings are sufficient to notify users of this behavior. Users can set warningfilter if they want to turn it into an error.